src/iop/colorequal.c

```diff
@@ -0,0 +1,1384 @@
+{
+  dt_iop_colorequal_data_t *d = (dt_iop_colorequal_data_t *)piece->data;
+
+  const int ch = piece->colors;
+  const float *const restrict in = (float*)i;
+  float *const restrict out = (float*)o;
+
+  const size_t npixels = (size_t)roi_out->width * roi_out->height;
+
+  // STEP 0: prepare the RGB <-> XYZ D65 matrices
+  // see colorbalancergb.c process() for the details, it's exactly the same
+  const struct dt_iop_order_iccprofile_info_t *const work_profile = dt_ioppr_get_pipe_current_profile_info(self, piece->pipe);
+  if(work_profile == NULL) return; // no point
+
+  dt_colormatrix_t input_matrix;
+  dt_colormatrix_t output_matrix;
+  dt_colormatrix_mul(input_matrix, XYZ_D50_to_D65_CAT16, work_profile->matrix_in);
+  dt_colormatrix_mul(output_matrix, work_profile->matrix_out, XYZ_D65_to_D50_CAT16);
+
+  float *const restrict U = dt_alloc_align_float(npixels);
+  float *const restrict V = dt_alloc_align_float(npixels);
+  float *const restrict corrections = dt_alloc_align_float(npixels * 4);
+
+  const float white = Y_to_dt_UCS_L_star(d->white_level);
+
+  // STEP 1: convert image from RGB to darktable UCS LUV then HSB
+#ifdef _OPENMP
+#pragma omp parallel for simd default(none) \
+  dt_omp_firstprivate(ch, npixels, in, out, U, V, corrections, input_matrix, d, white)  \
+  schedule(simd:static) aligned(in, out, U, V, corrections, input_matrix : 64)
+#endif
+  for(size_t k = 0; k < npixels; k++)
+  {
+    const float *const restrict pix_in = __builtin_assume_aligned(in + k * ch, 16);
+    float *const restrict pix_out = __builtin_assume_aligned(out + k * ch, 16);
+    float *const restrict corrections_out = __builtin_assume_aligned(corrections + k * ch, 16);
+
+    // Convert to XYZ D65
+    dt_aligned_pixel_t XYZ_D65 = { 0.f };
+    dot_product(pix_in, input_matrix, XYZ_D65);
+
+    // Convert to dt UCS 22 UV and store UV
+    dt_aligned_pixel_t xyY = { 0.f };
+    dt_XYZ_to_xyY(XYZ_D65, xyY);
+    float UV[2];
+    xyY_to_dt_UCS_UV(xyY, UV);
+    U[k] = UV[0];
+    V[k] = UV[1];
+
+    // Finish the conversion to dt UCS JCH then HSB
+    dt_aligned_pixel_t JCH = { 0.f };
+    dt_UCS_LUV_to_JCH(Y_to_dt_UCS_L_star(xyY[2]), white, UV, JCH);
+    dt_UCS_JCH_to_HSB(JCH, pix_out);
+
+    // Get the boosts
+    corrections_out[1] = lookup_gamut(d->LUT_saturation, pix_out[0]);
+    corrections_out[0] = lookup_gamut(d->LUT_hue, pix_out[0]);
+    corrections_out[2] = lookup_gamut(d->LUT_brightness, pix_out[0]);
+
+    // Copy alpha
+    pix_out[3] = pix_in[3];
+  }
+
+  // STEP 2: apply a guided filter on the corrections, guided with UV chromaticity, to ensure
+  // spatially-contiguous corrections even though the hue is not perfectly constant
+  // this will help avoiding chroma noise.
+  if(d->use_filter)
+    _guide_with_chromaticity(U, V, corrections, roi_out->width, roi_out->height, d->size, 1.f / d->feathering);
+
+  fprintf(stdout, "use filter: %i, feathering: %f, size: %f\n", d->use_filter, d->feathering, d->size);
```

Comment:
I understand this is draft, but just in case, this should be removed to turned into a dt_print controlled by a flag.